### PR TITLE
Local storage

### LINKED ISSUE #668: Cache downloaded metadata
@audax commented on [Wed Nov 25 2015](https://github.com/Qabel/qabel-android/issues/47)

Metadata should be served from the cache first and then updated. The cache should understand documentIDs.

### PR DESCRIPTION
**Pull request by JoGir**
resolves #668
moved box interactors to core-client module

### CODE REVIEW

File: box/src/main/java/de/qabel/box/storage/FolderNavigation.kt [hunk 1]
```diff
@@ -28,7 +28,7 @@ class FolderNavigation(
 
     @Throws(QblStorageException::class)
     override fun reloadMetadata(): DirectoryMetadata {
-        logger.trace("Reloading directory metadata")
+        logger.trace("Reloading directory metadata $path")
```
Comment (audax) [inline box/src/main/java/de/qabel/box/storage/FolderNavigation.kt]: What about paths with `%s` in them? Is our logger already "dumb"?

File: client/src/main/java/de/qabel/client/box/extensions.kt [hunk 2]
```diff
@@ -0,0 +1,9 @@
+package de.qabel.client.box
+
+import de.qabel.client.box.interactor.BrowserEntry
+import de.qabel.client.box.interactor.DownloadSource
+import java.io.ByteArrayInputStream
+
+fun ByteArray.toDownloadSource(entry: BrowserEntry.File)
```
Comment (audax) [inline client/src/main/java/de/qabel/client/box/extensions.kt]: shouldn't this be in a test source root?

File: client/src/main/java/de/qabel/client/box/storage/BoxLocalStorage.kt [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+
+class BoxLocalStorage(private val storageFolder: File,
+                      private val tmpFolder: File,
+                      private val cryptoUtils: CryptoUtils,
+                      private val repository: LocalStorageRepository) : LocalStorage, QabelLog {
+
+    override fun getBoxFile(path: BoxPath.File,
+                            boxFile: BoxFile): File? {
+        return identifier(path, boxFile).let {
+            debug("Get file $it")
```
Comment (audax) [inline client/src/main/java/de/qabel/client/box/storage/BoxLocalStorage.kt]: Add a test for specials chars in filenames, maybe they blow up the log function like we experienced on android

File: client/src/main/java/de/qabel/client/box/storage/BoxLocalStorage.kt [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+                        return readFile(tmp)
+                    } else {
+                        throw QblStorageNotFound("Invalid key")
+                    }
+                }
+                repository.delete(entry.id)
+            }
+            file.delete()
+            repository.delete(entry.id)
+        } catch (ex: EntityNotFoundException) {
```
Comment (audax) [inline client/src/main/java/de/qabel/client/box/storage/BoxLocalStorage.kt]: Does this catch have to span the entire block?

File: client/src/main/java/de/qabel/client/box/storage/repository/migrations/LSMigration1460997045Init.kt [hunk 5]
```diff
@@ -0,0 +1,25 @@
+package de.qabel.client.box.storage.repository.migrations
+
+import de.qabel.core.repository.sqlite.migration.AbstractMigration
+import java.sql.Connection
+
+class LSMigration1460997045Init(connection: Connection) : AbstractMigration(connection) {
```
Comment (audax) [inline client/src/main/java/de/qabel/client/box/storage/repository/migrations/LSMigration1460997045Init.kt]: Better name: LocalStorageMigration, no need to shorten it.

File: box/src/main/java/de/qabel/box/storage/FolderNavigation.kt (hunk 1 above)
Comment (JoGir) [inline box/src/main/java/de/qabel/box/storage/FolderNavigation.kt]: https://github.com/Qabel/qabel-android/blob/master/qabelbox/src/main/java/de/qabel/core/logging/AndroidLoggerWrapper.kt#L37

File: client/src/main/java/de/qabel/client/box/storage/BoxLocalStorage.kt (hunk 3 above)
Comment (JoGir) [inline client/src/main/java/de/qabel/client/box/storage/BoxLocalStorage.kt]: tested in BoxLocalStorageTest

**Pull request merged**